ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.util;
+
+import io.confluent.ksql.execution.codegen.TypeContext;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.expression.tree.LambdaFunctionCall;
+import io.confluent.ksql.function.GenericsUtil;
+import io.confluent.ksql.function.KsqlScalarFunction;
+import io.confluent.ksql.function.UdfFactory;
+import io.confluent.ksql.function.types.GenericType;
+import io.confluent.ksql.function.types.LambdaType;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlLambda;
+import io.confluent.ksql.schema.ksql.types.SqlLambdaResolved;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public final class FunctionArgumentsUtil {
+
+  private FunctionArgumentsUtil() {
+
+  }
+
+  public static class FunctionArgumentsAndContext {
+    private final List<TypeContext> contexts;
+    private final List<SqlArgument> arguments;
+    private final SqlType returnType;
+    private final KsqlScalarFunction function;
+
+    public FunctionArgumentsAndContext(
+        final List<TypeContext> contexts,
+        final List<SqlArgument> arguments,
+        final SqlType returnType,
+        final KsqlScalarFunction function
+    ) {
+      this.contexts = contexts;
+      this.arguments = arguments;
+      this.returnType = returnType;
+      this.function = function;
+    }
+
+    public List<TypeContext> getContexts() {
+      return contexts;
+    }
+
+    public List<SqlArgument> getArguments() {
+      return arguments;
+    }
+
+    public SqlType getReturnType() {
+      return returnType;
+    }
+
+    public KsqlScalarFunction getFunction() {
+      return function;
+    }
+  }
+
+  /**
+   * Given a function call node, we have to do a two pass processing of the 
```

Comment:
nit: explain what the function is doing first:

something like
"Compute type information given a function call node. Specifically, computes the function return type, the types of all arguments, and the types of all lambda parameters for arguments that are lambda expressions."